File: ecps-go/pkg/trust/jwt_rotation.go
// Package trust provides JWT secret rotation for ECPS-Go SDK
package trust

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// JWTSecret represents a JWT signing secret with metadata
type JWTSecret struct {
	Secret    string    `json:"secret"`
	CreatedAt time.Time `json:"created_at"`
	ExpiresAt time.Time `json:"expires_at"`
	KeyID     string    `json:"key_id"`
	Algorithm string    `json:"algorithm"`
}

// IsExpired checks if the secret has expired
func (s *JWTSecret) IsExpired() bool {
	return time.Now().UTC().After(s.ExpiresAt)
}

// IsNearExpiry checks if the secret is near expiry
func (s *JWTSecret) IsNearExpiry(thresholdHours int) bool {
	threshold := time.Now().UTC().Add(time.Duration(thresholdHours) * time.Hour)
	return s.ExpiresAt.Before(threshold)
}

// JWTSecretManager manages JWT secret rotation and validation
type JWTSecretManager struct {
	rotationIntervalHours int
	secretLength          int
	algorithm             string
	storagePath           string
	currentSecret         *JWTSecret
	previousSecret        *JWTSecret
	mutex                 sync.RWMutex
	stopChan              chan struct{}
	callbacks             map[string]func(*JWTSecret)
}

// NewJWTSecretManager creates a new JWT secret manager
func NewJWTSecretManager(rotationIntervalHours, secretLength int, algorithm, storagePath string) *JWTSecretManager {
	if storagePath == "" {
		homeDir, _ := os.UserHomeDir()
		storagePath = filepath.Join(homeDir, ".ecps", "jwt_secrets.json")
	}

	// Ensure storage directory exists
	os.MkdirAll(filepath.Dir(storagePath), 0700)

	return &JWTSecretManager{
		rotationIntervalHours: rotationIntervalHours,
		secretLength:          secretLength,
		algorithm:             algorithm,
		storagePath:           storagePath,
		stopChan:              make(chan struct{}),
		callbacks:             make(map[string]func(*JWTSecret)),
	}
}

// RegisterRotationCallback registers a callback for secret rotation
func (m *JWTSecretManager) RegisterRotationCallback(name string, callback func(*JWTSecret)) {
	m.mutex.Lock()
	defer m.mutex.Unlock()
	m.callbacks[name] = callback
	log.Printf("Registered rotation callback: %s", name)
}

// UnregisterRotationCallback unregisters a rotation callback
func (m *JWTSecretManager) UnregisterRotationCallback(name string) {
	m.mutex.Lock()
	defer m.mutex.Unlock()
	delete(m.callbacks, name)
	log.Printf("Unregistered rotation callback: %s", name)
}

// generateSecret generates a new JWT secret
func (m *JWTSecretManager) generateSecret() (*JWTSecret, error) {
	// Generate random bytes
	secretBytes := make([]byte, m.secretLength)
	if _, err := rand.Read(secretBytes); err != nil {
		return nil, fmt.Errorf("failed to generate secret: %w", err)
	}

	// Generate key ID
	keyIDBytes := make([]byte, 8)
	if _, err := rand.Read(keyIDBytes); err != nil {
		return nil, fmt.Errorf("failed to generate key ID: %w", err)
	}

	secret := base64.URLEncoding.EncodeToString(secretBytes)
	keyID := fmt.Sprintf("%x", keyIDBytes)

	now := time.Now().UTC()
	expiresAt := now.Add(time.Duration(m.rotationIntervalHours*2) * time.Hour) // Grace period

	jwtSecret := &JWTSecret{
		Secret:    secret,
		CreatedAt: now,
		ExpiresAt: expiresAt,
		KeyID:     keyID,
		Algorithm: m.algorithm,
	}

	log.Printf("Generated new JWT secret with key_id: %s", keyID)
	return jwtSecret, nil
}

// RotateSecretOnStartup rotates JWT secret on startup
func (m *JWTSecretManager) RotateSecretOnStartup() (*JWTSecret, error) {
	log.Println("Rotating JWT secret on startup")

	m.mutex.Lock()
	defer m.mutex.Unlock()

	// Load previous secrets if they exist
	if err := m.loadSecretsFromStorage(); err != nil {
		log.Printf("Warning: failed to load existing secrets: %v", err)
	}

	// Store current as previous
	if m.currentSecret != nil {
		m.previousSecret = m.currentSecret
		log.Printf("Moved current secret %s to previous", m.currentSecret.KeyID)
	}

	// Generate new current secret
	newSecret, err := m.generateSecret()
	if err != nil {
		return nil, fmt.Errorf("failed to generate new secret: %w", err)
	}
	m.currentSecret = newSecret

	// Save to storage
	if err := m.saveSecretsToStorage(); err != nil {
		log.Printf("Warning: failed to save secrets: %v", err)
	}

	// Notify callbacks
	m.notifyRotationCallbacks()

	log.Printf("JWT secret rotated on startup. New key_id: %s", newSecret.KeyID)
	return newSecret, nil
}

// StartAutomaticRotation starts automatic JWT secret rotation
func (m *JWTSecretManager) StartAutomaticRotation() {
	log.Printf("Starting automatic JWT secret rotation every %d hours", m.rotationIntervalHours)

	go func() {
		ticker := time.NewTicker(time.Duration(m.rotationIntervalHours) * time.Hour)
		defer ticker.Stop()

		for {
			select {
			case <-ticker.C:
				m.mutex.Lock()
				if m.currentSecret == nil || m.currentSecret.IsNearExpiry(m.rotationIntervalHours) {
					log.Println("Performing scheduled JWT secret rotation")
					if err := m.performRotation(); err != nil {
						log.Printf("Error in scheduled rotation: %v", err)
					}
				}
				m.mutex.Unlock()

			case <-m.stopChan:
				log.Println("JWT rotation loop stopped")
				return
			}
		}
	}()
}

// StopAutomaticRotation stops automatic JWT secret rotation
func (m *JWTSecretManager) StopAutomaticRotation() {
	close(m.stopChan)
	log.Println("Stopped automatic JWT secret rotation")
}

// performRotation performs secret rotation (must be called with lock held)
func (m *JWTSecretManager) performRotation() error {
	// Store current as previous
	if m.currentSecret != nil {
		m.previousSecret = m.currentSecret
		log.Printf("Moved current secret %s to previous", m.currentSecret.KeyID)
	}

	// Generate new current secret
	newSecret, err := m.generateSecret()
	if err != nil {
		return fmt.Errorf("failed to generate new secret: %w", err)
	}
	m.currentSecret = newSecret

	// Save to storage
	if err := m.saveSecretsToStorage(); err != nil {
		log.Printf("Warning: failed to save secrets: %v", err)
	}

	// Notify callbacks
	m.notifyRotationCallbacks()

	log.Printf("JWT secret rotated. New key_id: %s", newSecret.KeyID)
	return nil
}

// notifyRotationCallbacks notifies all registered callbacks
func (m *JWTSecretManager) notifyRotationCallbacks() {
	if m.currentSecret == nil {
		return
	}

	for name, callback := range m.callbacks {
		func() {
			defer func() {
				if r := recover(); r != nil {
					log.Printf("Error in rotation callback %s: %v", name, r)
				}
			}()
			callback(m.currentSecret)
			log.Printf("Notified rotation callback: %s", name)
		}()
	}
}

// GetCurrentSecret returns the current JWT secret
func (m *JWTSecretManager) GetCurrentSecret() *JWTSecret {
	m.mutex.RLock()
	defer m.mutex.RUnlock()
	return m.currentSecret
}

// GetPreviousSecret returns the previous JWT secret
func (m *JWTSecretManager) GetPreviousSecret() *JWTSecret {
	m.mutex.RLock()
	defer m.mutex.RUnlock()
	return m.previousSecret
}

// ValidateToken validates a JWT token using current or previous secret
func (m *JWTSecretManager) ValidateToken(tokenString string) (jwt.MapClaims, error) {
	m.mutex.RLock()
	defer m.mutex.RUnlock()

	var secretsToTry []*JWTSecret
	if m.currentSecret != nil {
		secretsToTry = append(secretsToTry, m.currentSecret)
	}
	if m.previousSecret != nil && !m.previousSecret.IsExpired() {
		secretsToTry = append(secretsToTry, m.previousSecret)
	}

	var lastErr error
	for _, jwtSecret := range secretsToTry {
		token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
			// Validate signing method
			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
				return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
			}
			return []byte(jwtSecret.Secret), nil
		})

		if err == nil && token.Valid {
			if claims, ok := token.Claims.(jwt.MapClaims); ok {
				log.Printf("Token validated with key_id: %s", jwtSecret.KeyID)
				return claims, nil
			}
		}
		lastErr = err
	}

	log.Println("JWT token validation failed with all available secrets")
	if lastErr != nil {
		return nil, lastErr
	}
	return nil, fmt.Errorf("no valid secrets available")
}

// CreateToken creates a JWT token with the current secret
func (m *JWTSecretManager) CreateToken(payload jwt.MapClaims, expiresInHours int) (string, error) {
	m.mutex.RLock()
	defer m.mutex.RUnlock()

	if m.currentSecret == nil {
		return "", fmt.Errorf("no current JWT secret available")
	}

	// Add standard claims
	now := time.Now().UTC()
	payload["iat"] = now.Unix()
	payload["exp"] = now.Add(time.Duration(expiresInHours) * time.Hour).Unix()
	payload["kid"] = m.currentSecret.KeyID

	// Create token
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, payload)
	tokenString, err := token.SignedString([]byte(m.currentSecret.Secret))
	if err != nil {
		return "", fmt.Errorf("failed to sign token: %w", err)
	}

	log.Printf("Created JWT token with key_id: %s", m.currentSecret.KeyID)
	return tokenString, nil
}

// saveSecretsToStorage saves secrets to persistent storage
func (m *JWTSecretManager) saveSecretsToStorage() error {
	data := map[string]interface{}{
		"updated_at": time.Now().UTC(),
	}

	if m.currentSecret != nil {
		data["current"] = m.currentSecret
	}
	if m.previousSecret != nil {
		data["previous"] = m.previousSecret
	}

	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal secrets: %w", err)
	}

	if err := ioutil.WriteFile(m.storagePath, jsonData, 0600); err != nil {
		return fmt.Errorf("failed to write secrets file: %w", err)
	}

	log.Printf("Saved JWT secrets to %s", m.storagePath)
	return nil
}

// loadSecretsFromStorage loads secrets from persistent storage
func (m *JWTSecretManager) loadSecretsFromStorage() error {
	if _, err := os.Stat(m.storagePath); os.IsNotExist(err) {
		log.Println("No existing JWT secrets found")
		return nil
	}

	data, err := ioutil.ReadFile(m.storagePath)
	if err != nil {
		return fmt.Errorf("failed to read secrets file: %w", err)
	}

	var secretsData map[string]interface{}
	if err := json.Unmarshal(data, &secretsData); err != nil {
		return fmt.Errorf("failed to unmarshal secrets: %w", err)
	}

	if currentData, ok := secretsData["current"]; ok {
		currentBytes, _ := json.Marshal(currentData)
		var current JWTSecret
		if err := json.Unmarshal(currentBytes, &current); err == nil {
			m.currentSecret = &current
			log.Printf("Loaded current secret with key_id: %s", current.KeyID)
		}
	}

	if previousData, ok := secretsData["previous"]; ok {
		previousBytes, _ := json.Marshal(previousData)
		var previous JWTSecret
		if err := json.Unmarshal(previousBytes, &previous); err == nil {
			m.previousSecret = &previous
			log.Printf("Loaded previous secret with key_id: %s", previous.KeyID)
		}
	}

	return nil
}

// Global JWT secret manager instance
var globalJWTManager *JWTSecretManager

// InitializeJWTRotation initializes JWT secret rotation on startup
func InitializeJWTRotation(rotationIntervalHours, secretLength int, algorithm, storagePath string) (*JWTSecret, error) {
	globalJWTManager = NewJWTSecretManager(rotationIntervalHours, secretLength, algorithm, storagePath)
	return globalJWTManager.RotateSecretOnStartup()
}

// GetJWTManager returns the global JWT secret manager instance
func GetJWTManager() *JWTSecretManager {
	return globalJWTManager
}

// StartJWTRotation starts automatic JWT secret rotation
func StartJWTRotation() {
	if globalJWTManager != nil {
		globalJWTManager.StartAutomaticRotation()
	}
}

// StopJWTRotation stops automatic JWT secret rotation
func StopJWTRotation() {
	if globalJWTManager != nil {
		globalJWTManager.StopAutomaticRotation()
	}
}

File: ecps_uv/trust/jwt_rotation.py
"""
JWT Secret Rotation for ECPS-UV SDK.

This module provides automatic JWT secret rotation on startup and periodic rotation
to enhance security for distributed robotic systems.
"""

import asyncio
import logging
import os
import secrets
import time
from datetime import datetime, timedelta
from typing import Dict, Optional, Callable, Any
from dataclasses import dataclass
from threading import Lock
import jwt
import json

logger = logging.getLogger("ecps_uv.trust.jwt_rotation")


@dataclass
class JWTSecret:
    """JWT secret with metadata."""
    secret: str
    created_at: datetime
    expires_at: datetime
    key_id: str
    algorithm: str = "HS256"
    
    def is_expired(self) -> bool:
        """Check if the secret has expired."""
        return datetime.utcnow() > self.expires_at
    
    def is_near_expiry(self, threshold_hours: int = 24) -> bool:
        """Check if the secret is near expiry."""
        threshold = datetime.utcnow() + timedelta(hours=threshold_hours)
        return self.expires_at <= threshold


class JWTSecretManager:
    """Manages JWT secret rotation and validation."""
    
    def __init__(
        self,
        rotation_interval_hours: int = 24,
        secret_length: int = 64,
        algorithm: str = "HS256",
        storage_path: Optional[str] = None
    ):
        """
        Initialize JWT secret manager.
        
        Args:
            rotation_interval_hours: Hours between secret rotations
            secret_length: Length of generated secrets in bytes
            algorithm: JWT signing algorithm
            storage_path: Path to store secret metadata (optional)
        """
        self.rotation_interval_hours = rotation_interval_hours
        self.secret_length = secret_length
        self.algorithm = algorithm
        self.storage_path = storage_path or os.path.join(
            os.path.expanduser("~"), ".ecps", "jwt_secrets.json"
        )
        
        self._current_secret: Optional[JWTSecret] = None
        self._previous_secret: Optional[JWTSecret] = None
        self._lock = Lock()
        self._rotation_task: Optional[asyncio.Task] = None
        self._callbacks: Dict[str, Callable[[JWTSecret], None]] = {}
        
        # Ensure storage directory exists
        os.makedirs(os.path.dirname(self.storage_path), exist_ok=True)
    
    def register_rotation_callback(self, name: str, callback: Callable[[JWTSecret], None]) -> None:
        """
        Register a callback to be called when secrets are rotated.
        
        Args:
            name: Unique name for the callback
            callback: Function to call with new secret
        """
        self._callbacks[name] = callback
        logger.debug(f"Registered rotation callback: {name}")
    
    def unregister_rotation_callback(self, name: str) -> None:
        """Unregister a rotation callback."""
        self._callbacks.pop(name, None)
        logger.debug(f"Unregistered rotation callback: {name}")
    
    def generate_secret(self) -> JWTSecret:
        """Generate a new JWT secret."""
        secret_bytes = secrets.token_bytes(self.secret_length)
        secret = secrets.token_urlsafe(self.secret_length)
        key_id = secrets.token_hex(8)
        
        now = datetime.utcnow()
        expires_at = now + timedelta(hours=self.rotation_interval_hours * 2)  # Grace period
        
        jwt_secret = JWTSecret(
            secret=secret,
            created_at=now,
            expires_at=expires_at,
            key_id=key_id,
            algorithm=self.algorithm
        )
        
        logger.info(f"Generated new JWT secret with key_id: {key_id}")
        return jwt_secret
    
    def rotate_secret_on_startup(self) -> JWTSecret:
        """
        Rotate JWT secret on startup.
        
        Returns:
            The new JWT secret
        """
        logger.info("Rotating JWT secret on startup")
        
        with self._lock:
            # Load previous secret if exists
            self._load_secrets_from_storage()
            
            # Store current as previous
            if self._current_secret:
                self._previous_secret = self._current_secret
                logger.debug(f"Moved current secret {self._current_secret.key_id} to previous")
            
            # Generate new current secret
            self._current_secret = self.generate_secret()
            
            # Save to storage
            self._save_secrets_to_storage()
            
            # Notify callbacks
            self._notify_rotation_callbacks()
            
            logger.info(f"JWT secret rotated on startup. New key_id: {self._current_secret.key_id}")
            return self._current_secret
    
    async def start_automatic_rotation(self) -> None:
        """Start automatic JWT secret rotation."""
        if self._rotation_task and not self._rotation_task.done():
            logger.warning("Automatic rotation already running")
            return
        
        logger.info(f"Starting automatic JWT secret rotation every {self.rotation_interval_hours} hours")
        self._rotation_task = asyncio.create_task(self._rotation_loop())
    
    async def stop_automatic_rotation(self) -> None:
        """Stop automatic JWT secret rotation."""
        if self._rotation_task:
            self._rotation_task.cancel()
            try:
                await self._rotation_task
            except asyncio.CancelledError:
                pass
            self._rotation_task = None
            logger.info("Stopped automatic JWT secret rotation")
    
    async def _rotation_loop(self) -> None:
        """Main rotation loop."""
        while True:
            try:
                # Wait for rotation interval
                await asyncio.sleep(self.rotation_interval_hours * 3600)
                
                # Check if rotation is needed
                with self._lock:
                    if (self._current_secret is None or 
                        self._current_secret.is_near_expiry(threshold_hours=self.rotation_interval_hours)):
                        
                        logger.info("Performing scheduled JWT secret rotation")
                        self._perform_rotation()
                
            except asyncio.CancelledError:
                logger.info("JWT rotation loop cancelled")
                break
            except Exception as e:
                logger.error(f"Error in JWT rotation loop: {e}")
                # Continue the loop after error
                await asyncio.sleep(60)  # Wait 1 minute before retrying
    
    def _perform_rotation(self) -> None:
        """Perform secret rotation (must be called with lock held)."""
        # Store current as previous
        if self._current_secret:
            self._previous_secret = self._current_secret
            logger.debug(f"Moved current secret {self._current_secret.key_id} to previous")
        
        # Generate new current secret
        self._current_secret = self.generate_secret()
        
        # Save to storage
        self._save_secrets_to_storage()
        
        # Notify callbacks
        self._notify_rotation_callbacks()
        
        logger.info(f"JWT secret rotated. New key_id: {self._current_secret.key_id}")
    
    def _notify_rotation_callbacks(self) -> None:
        """Notify all registered callbacks of secret rotation."""
        if not self._current_secret:
            return
        
        for name, callback in self._callbacks.items():
            try:
                callback(self._current_secret)
                logger.debug(f"Notified rotation callback: {name}")
            except Exception as e:
                logger.error(f"Error in rotation callback {name}: {e}")
    
    def get_current_secret(self) -> Optional[JWTSecret]:
        """Get the current JWT secret."""
        with self._lock:
            return self._current_secret
    
    def get_previous_secret(self) -> Optional[JWTSecret]:
        """Get the previous JWT secret (for validation during transition)."""
        with self._lock:
            return self._previous_secret
    
    def validate_token(self, token: str) -> Dict[str, Any]:
        """
        Validate a JWT token using current or previous secret.
        
        Args:
            token: JWT token to validate
            
        Returns:
            Decoded token payload
            
        Raises:
            jwt.InvalidTokenError: If token is invalid
        """
        with self._lock:
            secrets_to_try = []
            
            if self._current_secret:
                secrets_to_try.append(self._current_secret)
            
            if self._previous_secret and not self._previous_secret.is_expired():
                secrets_to_try.append(self._previous_secret)
        
        last_error = None
        for jwt_secret in secrets_to_try:
            try:
                payload = jwt.decode(
                    token,
                    jwt_secret.secret,
                    algorithms=[jwt_secret.algorithm]
                )
                logger.debug(f"Token validated with key_id: {jwt_secret.key_id}")
                return payload
            except jwt.InvalidTokenError as e:
                last_error = e
                continue
        
        # If we get here, validation failed with all secrets
        logger.warning("JWT token validation failed with all available secrets")
        raise last_error or jwt.InvalidTokenError("No valid secrets available")
    
    def create_token(self, payload: Dict[str, Any], expires_in_hours: int = 1) -> str:
        """
        Create a JWT token with the current secret.
        
        Args:
            payload: Token payload
            expires_in_hours: Token expiration time in hours
            
        Returns:
            Encoded JWT token
        """
        with self._lock:
            if not self._current_secret:
                raise RuntimeError("No current JWT secret available")
            
            # Add standard claims
            now = datetime.utcnow()
            payload.update({
                'iat': now,
                'exp': now + timedelta(hours=expires_in_hours),
                'kid': self._current_secret.key_id
            })
            
            token = jwt.encode(
                payload,
                self._current_secret.secret,
                algorithm=self._current_secret.algorithm
            )
            
            logger.debug(f"Created JWT token with key_id: {self._current_secret.key_id}")
            return token
    
    def _save_secrets_to_storage(self) -> None:
        """Save secrets to persistent storage."""
        try:
            data = {
                'current': self._secret_to_dict(self._current_secret) if self._current_secret else None,
                'previous': self._secret_to_dict(self._previous_secret) if self._previous_secret else None,
                'updated_at': datetime.utcnow().isoformat()
            }
            
            with open(self.storage_path, 'w') as f:
                json.dump(data, f, indent=2)
            
            # Set restrictive permissions
            os.chmod(self.storage_path, 0o600)
            logger.debug(f"Saved JWT secrets to {self.storage_path}")
            
        except Exception as e:
            logger.error(f"Failed to save JWT secrets: {e}")
    
    def _load_secrets_from_storage(self) -> None:
        """Load secrets from persistent storage."""
        try:
            if not os.path.exists(self.storage_path):
                logger.debug("No existing JWT secrets found")
                return
            
            with open(self.storage_path, 'r') as f:
                data = json.load(f)
            
            if data.get('current'):
                self._current_secret = self._dict_to_secret(data['current'])
                logger.debug(f"Loaded current secret with key_id: {self._current_secret.key_id}")
            
            if data.get('previous'):
                self._previous_secret = self._dict_to_secret(data['previous'])
                logger.debug(f"Loaded previous secret with key_id: {self._previous_secret.key_id}")
                
        except Exception as e:
            logger.error(f"Failed to load JWT secrets: {e}")
    
    def _secret_to_dict(self, secret: JWTSecret) -> Dict[str, Any]:
        """Convert JWTSecret to dictionary for storage."""
        return {
            'secret': secret.secret,
            'created_at': secret.created_at.isoformat(),
            'expires_at': secret.expires_at.isoformat(),
            'key_id': secret.key_id,
            'algorithm': secret.algorithm
        }
    
    def _dict_to_secret(self, data: Dict[str, Any]) -> JWTSecret:
        """Convert dictionary to JWTSecret."""
        return JWTSecret(
            secret=data['secret'],
            created_at=datetime.fromisoformat(data['created_at']),
            expires_at=datetime.fromisoformat(data['expires_at']),
            key_id=data['key_id'],
            algorithm=data['algorithm']
        )


# Global JWT secret manager instance
_jwt_manager: Optional[JWTSecretManager] = None


def get_jwt_manager() -> JWTSecretManager:
    """Get the global JWT secret manager instance."""
    global _jwt_manager
    if _jwt_manager is None:
        _jwt_manager = JWTSecretManager()
    return _jwt_manager


def initialize_jwt_rotation(
    rotation_interval_hours: int = 24,
    secret_length: int = 64,
    algorithm: str = "HS256",
    storage_path: Optional[str] = None
) -> JWTSecret:
    """
    Initialize JWT secret rotation on startup.
    
    Args:
        rotation_interval_hours: Hours between secret rotations
        secret_length: Length of generated secrets in bytes
        algorithm: JWT signing algorithm
        storage_path: Path to store secret metadata
        
    Returns:
        The initial JWT secret
    """
    global _jwt_manager
    _jwt_manager = JWTSecretManager(
        rotation_interval_hours=rotation_interval_hours,
        secret_length=secret_length,
        algorithm=algorithm,
        storage_path=storage_path
    )
    
    return _jwt_manager.rotate_secret_on_startup()


async def start_jwt_rotation() -> None:
    """Start automatic JWT secret rotation."""
    manager = get_jwt_manager()
    await manager.start_automatic_rotation()


async def stop_jwt_rotation() -> None:
    """Stop automatic JWT secret rotation."""
    manager = get_jwt_manager()
    await manager.stop_automatic_rotation()